keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,191 @@
+---
+title: KEP Template
+authors:
+  - "@jpbetz"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-scalability
+reviewers:
+  - TBD
+  - "@wojtek-t"
+  - "@jingyih"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-12-10
+last-updated: 2019-12-10
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Consistent Reads from Cache
+
+Kubernetes Get and List requests are guaranteed to be "consistent reads" if the
+resourceVerions paramter is not provided and are served from etcd using a
+"quorum read".
+
+But often the watch cache contains sufficiently up-to-date data to serve the
+read request, and could serve it far more efficiently.
+
+This KEP proposes a mechanism to serve most reads from the watch cache
+while still providing the same consistency guarantees as serving the
+read from etcd.
+
+## Table of Contents
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Rollout Plan](#rollout-plan)
+    - [Reflector](#reflector)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Consistent reads may be served from cache so long as:
+- A consistent (quorum) read is first made to etcd to get the latest "revsion"
+- The data in the watch cache no older than the latest "revsion" just from etcd
+
+etcd 3.4 contains a [progress notify](https://github.com/etcd-io/etcd/issues/9855)
+feature that was added specifically to make it easy to check if a cache that is
+updated via an etcd watch is up-to-date relative to some etcd revsion.
+
+This KEP summarizes how we can take advantage of progress notify feature in etcd
+3.4 to efficiently determine how up-to-date kubernetes watch caches are then
+serve reads from the watch cache when they are sufficiently up-to-date.
+
+## Motivation
+
+Intuitively, we expect that serving reads from the watch cache to be more
```

Comment:
It's not just intuitively - it is based on data and facts.
If serving from etcd, all data has to be send from etcd and deserialized in apiserver. Those steps doesn't exist when listing from watch cache.

Reply:
Absolutely. I've firmed this up.